@arnova Sorry to bother you again :-)

a) new warning
```
13:19:08 /home/jenkins/workspace/LINUX-64/xbmc/filesystem/FileCache.cpp:318:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```
b) https://jenkins.kodi.tv/job/BuildMulti-PR/14207/artifact/PR17081.diff

I hope we get it done in 2k19 :P

--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -90,7 +90,6 @@ CFileCache::CFileCache(const unsigned int flags)
   , m_writeRate(0)
   , m_writeRateActual(0)
   , m_forwardCacheSize(0)
-  , m_forward(0)
   , m_bFilling(false)
   , m_bLowSpeedDetected(false)
   , m_fileSize(0)
@@ -200,7 +199,6 @@ bool CFileCache::Open(const CURL& url)
   m_writePos = 0;
   m_writeRate = 1024 * 1024;
   m_writeRateActual = 0;
-  m_forward = 0;
   m_bFilling = true;
   m_bLowSpeedDetected = false;
   m_seekEvent.Reset();
@@ -268,7 +266,6 @@ void CFileCache::Process()
           CLog::Log(LOGDEBUG,
                     "CFileCache::Process - Cache completely reset for seek to position %" PRId64,
                     m_seekPos);
-          m_forward = 0;
           m_bFilling = true;
           m_bLowSpeedDetected = false;
         }
@@ -400,11 +397,11 @@ void CFileCache::Process()
     // avoid uncertainty at start of caching
     m_writeRateActual = average.Rate(m_writePos, 1000);
 
-    // Update forward cache size
-    m_forward = m_pCache->WaitForData(0, 0);
-
     // NOTE: Hysteresis (20-80%) for filling-logic
-    const float level = (m_forwardCacheSize == 0) ? 0.0 : (float) m_forward / m_forwardCacheSize;
+    const int64_t forward = m_pCache->WaitForData(0, 0);
+    const float level =
+        (m_forwardCacheSize == 0) ? 0.0 : static_cast<float>(forward / m_forwardCacheSize);
+
     if (level > 0.8f)
     {
      /* NOTE: We can only reliably test for low speed condition, when the cache is *really*
@@ -589,7 +586,7 @@ int CFileCache::IoControl(EIoControl request, void* param)
   if (request == IOCTRL_CACHE_STATUS)
   {
     SCacheStatus* status = (SCacheStatus*)param;
-    status->forward = m_forward;
+    status->forward = m_pCache->WaitForData(0, 0);
     status->maxrate = m_writeRate;
     status->currate = m_writeRateActual;
     status->lowspeed = m_bLowSpeedDetected;

--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -155,27 +155,40 @@ bool CFileCache::Open(const CURL& url)
       size_t cacheSize;
       if (m_fileSize > 0 && m_fileSize < CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheMemSize && !(m_flags & READ_AUDIO_VIDEO))
       {
-        // NOTE: We don't need to take into account READ_MULTI_STREAM here as it's only used for audio/video
+        // Cap cache size by filesize, but not for audio/video files as those may grow.
+        // We don't need to take into account READ_MULTI_STREAM here as that's only used for audio/video
         cacheSize = m_fileSize;
+
+        // Cap chunk size by cache size
+        if (m_chunkSize > cacheSize)
+          m_chunkSize = cacheSize;
       }
       else
       {
         cacheSize = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheMemSize;
+
+        // NOTE: READ_MULTI_STREAM is only used with READ_AUDIO_VIDEO
+        if (m_flags & READ_MULTI_STREAM)
+        {
+          // READ_MULTI_STREAM requires double buffering, so use half the amount of memory for each buffer
+          cacheSize /= 2;
+        }
+
+        // Make sure cache can at least hold 2 chunks
+        if (cacheSize < m_chunkSize * 2)
+          cacheSize = m_chunkSize * 2;
       }
-
-      // Cap chunk size by cache size
-      if (m_chunkSize > cacheSize)
-        m_chunkSize = cacheSize;
-
-      size_t back = cacheSize / 4;
-      size_t front = cacheSize - back;
 
       if (m_flags & READ_MULTI_STREAM)
-      {
-        // READ_MULTI_STREAM requires double buffering, so use half the amount of memory for each buffer
-        front /= 2;
-        back /= 2;
-      }
+        CLog::Log(LOGDEBUG, "CFileCache::Open - Using double memory cache each sized %i bytes",
+                  cacheSize);
+      else
+        CLog::Log(LOGDEBUG, "CFileCache::Open - Using single memory cache sized %i bytes",
+                  cacheSize);
+
+      const size_t back = cacheSize / 4;
+      const size_t front = cacheSize - back;
+
       m_pCache = std::unique_ptr<CCircularCache>(new CCircularCache(front, back)); // C++14 - Replace with std::make_unique
       m_forwardCacheSize = front;
     }

--- a/xbmc/filesystem/FileCache.cpp
+++ b/xbmc/filesystem/FileCache.cpp
@@ -164,6 +164,10 @@ bool CFileCache::Open(const CURL& url)
         cacheSize = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheMemSize;
       }
 
+      // Cap chunk size by cache size
+      if (m_chunkSize > cacheSize)
+        m_chunkSize = cacheSize;
+
       size_t back = cacheSize / 4;
       size_t front = cacheSize - back;
 
@@ -292,20 +296,25 @@ void CFileCache::Process()
       }
     }
 
-    size_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
+    const int64_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
+    int64_t maxSourceRead = m_chunkSize;
+    // Cap source read size by space available between current write position and EOF
+    if (m_fileSize != 0)
+      maxSourceRead = std::min(maxSourceRead, m_fileSize - m_writePos);
 
     /* Only read from source if there's enough write space in the cache
      * else we may keep disposing data and seeking back on (slow) source
      */
-    if (maxWrite < m_chunkSize && !cacheReachEOF)
+    if (maxWrite < maxSourceRead && !cacheReachEOF)
     {
+      // Wait until sufficient cache write space is available
       m_pCache->m_space.WaitMSec(5);
       continue;
     }
 
     ssize_t iRead = 0;
     if (!cacheReachEOF)
-      iRead = m_source.Read(buffer.get(), m_chunkSize);
+      iRead = m_source.Read(buffer.get(), maxSourceRead);
     if (iRead == 0)
     {
       // Check for actual EOF and retry as long as we still have data in our cache

--- a/xbmc/filesystem/FileCache.h
+++ b/xbmc/filesystem/FileCache.h
@@ -69,7 +69,6 @@ namespace XFILE
     unsigned m_writeRate;
     unsigned m_writeRateActual;
     int64_t m_forwardCacheSize;
-    int64_t m_forward;
     bool m_bFilling;
     bool m_bLowSpeedDetected;
     std::atomic<int64_t> m_fileSize;

